### DISCUSSION TITLE
Exclude polygons loaded at the build phase

### DISCUSSION THREAD

**Discussion by cdbattags** [valhalla/valhalla#4225] [Ideas]
There is currently a system for avoiding/excluding polygons in a route through `exclude_polygons`.

The issue is that it's per-request and seems to be quite a strain on the service?

Would we be open to adding a layer to the build phase that has a map of these avoid polygons by label? The service would then be able to lookup the polygon by name and cache whenever it's needed.

Use case:
- avoiding major cities
- avoid specific states

The nice part about this is that you'd be able to load any polygon(s) you wanted.

For reference:
- https://andrewvanleuven.com/post/cityshapefiles/

**Comment by cdbattags** [comment]
The lookup would happen somewhere near here:

https://github.com/valhalla/valhalla/blob/master/src/worker.cc#L1010C7-L1010C7

**Reply by nilsnolde** [reply]
I'd personally vote for pre-processing instead of including such pretty generic functionality in a single project, i.e. edit the PBF to throw out those ways you don't want before giving it to Valhalla. `osmium` already does what you want out-of-the-box with polygons and even multipolygons: https://docs.osmcode.org/osmium/latest/osmium-extract.html. Be aware it needs quite some memory for routing-compatible strategy.

If you need smth more custom (e.g. only throw certain tags like `highway=pedestrian` or other strategies), there's nothing very flexible off-the-shelf AFAIK (I might stand corrected by others e.g. @ImreSamu ?), so there'd be a space to fill in OSS I think:) If there's none, creating a generic/flexible new piece of software to pre-process PBFs for a router is smth that could be very valuable (with libosmium or one of its bindings), also for other things.

I was thinking about smth similar, just more in the direction of "soft-converting" OSM data to make it easier for routing engines to consume and to make them more consistent across the board, at least provide the best `maxspeed` via [osm-legal-default-speeds](https://github.com/westnordost/osm-legal-default-speeds) and `maxspeed:practical` via [OpenStreetMapSpeeds](https://github.com/OpenStreetMapSpeeds/schema/blob/master/schema.json) we can get, or maybe even getting rid of some relations (EDIT: elevation & density could be other important ones). There's some threads in our issues/discussions and over at OSRM's. I didn't find the time (and motivation) yet to investigate that path further, but would like to at some point. I might balk at rust though :smile:

**Reply by nilsnolde** [reply]
Actually it could be interesting to also have such an executable in Valhalla. Never had that use case before, but I guess it could be somewhat interesting to allow removing/nulling nodes/edges in an existing graph, which could be quite a quick thing to do (computationally) and wouldn't require a full graph re-build. But I'm pretty sure hardly any of the service runtime code is currently robust to just having empty/null edges somewhere in the middle, and it's maybe very much not advisable to do so. In any case, any such graph edit would require a service restart with all of the consequences.

But of course, you're right, `avoid_polygons` can be _very_ expensive on each request for large and/or complex polygons. It's more meant for construction site or similarly small areas.

**Reply by ImreSamu** [reply]
I agree... ( with @nilsnolde  ) 
if a quick solution is needed, then preprocessing (i.e., modifying the OSM graph) is the most ideal, 
which can be done in several ways and besides the mentioned osmium CLI tool, 
I would also recommend the [pyosmium](https://osmcode.org/pyosmium/) solution based on Python.

Regardless, the problem is interesting...  
it reminds me of the Ulysses problem ~ "cross Dublin without passing a pub". 
( https://googlemapsmania.blogspot.com/2014/06/happy-bloomsday-maps.html )
Its solution could be possible in several ways, for instance, the new Valhalla Landmark database could contain circular polygons with a radius of 50 meters around the pubs in Dublin, which the route finder should avoid.

In other words, it would be worthwhile to generalize the problem and not just implement avoidance for "major cities" and "states"...
but also for other freely chosen objects, such as: pubs, schools, railway crossings, police stations, ... to avoid.

**Reply by nilsnolde** [reply]
Oops haha I only now understand what you're really suggesting.. You wanna cache all the edge id's which are inside any of the polygons which the service reads during startup, based on static GeoJSON(s). That makes much more sense than what I was thinking about. I really like that idea. It'd would be much faster than currently at request time which honestly bothers me quite a bit. Changing any polygon would still need a service restart, but that's unavoidable. A few thoughts:

- we add a member to `loki_worker_t` which is set to `true` if there's a valid GeoJSON file path with polygons in the valhalla.json (new config `exclude_geojson` or so)
- at that point we find all the edge IDs which are falling inside those polygons and put them in the same spot in the PBF as the request `exclude_polygons` (the search algorithm can also be re-used!), so costing can use them.

That'd not be too much work, but I'm not sure how feasible it is to put so much stuff in the PBF. Maybe someone else has a better gut feeling about it than me. If it's 100s of 1000s or even millions, it might be a drag.. Definitely quicker than attempting the same currently.

WDYT @kevinkreiser ? I'd find it an interesting approach and it'd be perfectly compatible with the current `exclude_polygons`.

**Reply by nilsnolde** [reply]
The best alternative would actually be to use Valhalla's live traffic infrastructure and have a separate process update the traffic.tar's edge speeds to 0 to close them. No service restart, framework of your choice, and pretty much no performance difference. But again, not portable to other software like geocoders or other routers, that'll only be possible when manipulating the PBF directly.

Quite a few options to pick from:)
